src/FavIconManager.vala

```diff
@@ -36,15 +38,19 @@ public class FeedReader.FavIconManager : GLib.Object {
 		{
 			var feed_id = feed.getFeedID();
 			var future = m_map.get(feed_id);
-			if(future == null)
+
+			bool download_icon = future == null || (future.ready && (future.value.Metadata == null || future.value.Metadata.IsExpired()) );
+
+			if(download_icon)
 			{
-				var promise = new Gee.Promise<Gdk.Pixbuf?>();
+				var promise = new Gee.Promise<FavIconData?>();
 				try
 				{
 					future = promise.future;
 					m_map.set(feed_id, future);
 
-					var stream = yield downloadFavIcon(feed);
+					ResourceMetadata metadata;
+					var stream = yield downloadFavIcon(feed, out metadata);
```

Comment:
This is only going to get the current metadata (what was  on the disk when the sync started. The new metadata isn't returned from `downloadFavIcon()` :(